featurebyte/service/observation_table.py

```diff
@@ -359,27 +363,77 @@ async def update_observation_table(
         ------
         ObservationTableInvalidContextError
             If the entity ids are different from the existing Context
+        ObservationTableInvalidUseCaseError
+            If the use case is invalid
         """
 
+        if data.use_case_ids:
+            raise ObservationTableInvalidUseCaseError("use_case_ids is not a valid field to update")
+
+        observation_table = await self.get_document(document_id=observation_table_id)
+        if data.use_case_id_to_add or data.use_case_id_to_remove:
+            if not observation_table.context_id:
+                raise ObservationTableInvalidUseCaseError(
+                    f"Cannot add/remove UseCase as the ObservationTable {observation_table_id} is not associated with any Context."
+                )
+
+            use_case_ids = observation_table.use_case_ids
+            # validate use case id to add
+            if data.use_case_id_to_add:
+                if data.use_case_id_to_add in use_case_ids:
+                    raise ObservationTableInvalidUseCaseError(
+                        f"Cannot add UseCase {data.use_case_id_to_add} as it is already associated with the ObservationTable."
+                    )
+                use_case = await self.use_case_service.get_document(
+                    document_id=data.use_case_id_to_add
+                )
+                if use_case.context_id != observation_table.context_id:
+                    raise ObservationTableInvalidUseCaseError(
+                        f"Cannot add UseCase {data.use_case_id_to_add} as its context_id is different from the existing context_id."
+                    )
+
+                # validate request_input and target_id
+                if not isinstance(observation_table.request_input, TargetInput):
+                    raise ObservationTableInvalidUseCaseError(
+                        "observation table request_input is not TargetInput"
+                    )
+
+                if not use_case.target_id or (
+                    isinstance(observation_table.request_input, TargetInput)
+                    and observation_table.request_input.target_id != use_case.target_id
+                ):
+                    raise ObservationTableInvalidUseCaseError(
+                        f"Cannot add UseCase {data.use_case_id_to_add} as its target_id is different from the existing target_id."
+                    )
+
+                use_case_ids.append(data.use_case_id_to_add)
+
+            # validate use case id to remove
+            if data.use_case_id_to_remove:
+                if data.use_case_id_to_remove not in use_case_ids:
+                    raise ObservationTableInvalidUseCaseError(
+                        f"Cannot remove UseCase {data.use_case_id_to_remove} as it is not associated with the ObservationTable."
+                    )
```

Comment:
Would be good to have an unit test to cover this.

Reply:
I think the test is already in this PR: https://github.com/featurebyte/featurebyte/blob/yanhui/DEV-2435_2/tests/unit/routes/test_observation_table.py#L244